# Project: Category prediction of IKEA furniture

## Data Visualization

In [13]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt

%matplotlib inline
plt.style.use('ggplot') # emulate pretty r-style plots

In [14]:
#read the data and delete the index of entry
IKEA_df = pd.read_csv('IKEA_SA_Furniture_Web_Scrapings_sss.csv')
IKEA_df = IKEA_df.drop(IKEA_df.columns[0],axis =1)

In [15]:
IKEA_df.shape

(3694, 13)

We have 3694 samples and 12 features together with target catrgory as follows:

In [16]:
# data type  
IKEA_df.dtypes

item_id                int64
name                  object
category              object
price                float64
old_price             object
sellable_online         bool
link                  object
other_colors          object
short_description     object
designer              object
depth                float64
height               float64
width                float64
dtype: object

In [17]:
IKEA_df.count()

item_id              3694
name                 3694
category             3694
price                3694
old_price            3694
sellable_online      3694
link                 3694
other_colors         3694
short_description    3694
designer             3694
depth                2231
height               2706
width                3105
dtype: int64

We can see some data is missing in feature depth, height and width 

In [18]:
# maybe make this into a pretty pie/bar chart
IKEA_df.describe().round(2)

item_id    price    depth   height    width
count      3694.00  3694.00  2231.00  2706.00  3105.00
mean   48632396.79  1078.21    54.38   101.68   104.47
std    28887094.10  1374.65    29.96    61.10    71.13
min       58487.00     3.00     1.00     1.00     1.00
25%    20390574.00   180.90    38.00    67.00    60.00
50%    49288078.00   544.70    47.00    83.00    80.00
75%    70403572.75  1429.50    60.00   124.00   140.00
max    99932615.00  9585.00   257.00   700.00   420.00

In [19]:
nun =  IKEA_df['category'].nunique()
print(IKEA_df['category'].unique())
print('There are ' + str(nun) + ' categories in total' )

['Bar furniture' 'Beds' 'Bookcases & shelving units'
 'Cabinets & cupboards' 'Café furniture' 'Chairs'
 'Chests of drawers & drawer units' "Children's furniture"
 'Nursery furniture' 'Outdoor furniture' 'Room dividers'
 'Sideboards, buffets & console tables' 'Sofas & armchairs'
 'Tables & desks' 'Trolleys' 'TV & media furniture' 'Wardrobes']
There are 17 categories in total


In [20]:
nun =  IKEA_df['designer'].nunique()
#print(IKEA_df['designer'].unique())
print('There are ' + str(nun) + ' designers in total' )

There are 381 designers in total


In [21]:
IKEA_df.groupby(['designer','category'])['item_id'].count().reset_index(name='count') \
                             .sort_values(['count'], ascending=False) \
                             .head(10)

designer                    category  count
374                  IKEA of Sweden  Bookcases & shelving units    256
375                  IKEA of Sweden        Cabinets & cupboards     96
386                  IKEA of Sweden              Tables & desks     89
385                  IKEA of Sweden        TV & media furniture     78
259                 Ehlén Johansson           Sofas & armchairs     74
377                  IKEA of Sweden                      Chairs     68
427  IKEA of Sweden/Ehlén Johansson                   Wardrobes     62
273  Ehlén Johansson/IKEA of Sweden                   Wardrobes     61
307               Francis Cayouette           Sofas & armchairs     53
753                    Ola Wihlborg           Sofas & armchairs     48

In [22]:
IKEA_df['designer_len'] = IKEA_df['designer'].str.len() 
IKEA_df['designer_len'].describe()

count    3694.000000
mean       33.755820
std        86.016629
min         3.000000
25%        14.000000
50%        15.000000
75%        28.000000
max      1261.000000
Name: designer_len, dtype: float64

In [23]:
IKEA_df[IKEA_df['designer_len'] > 50].shape

(166, 14)

In [24]:
IKEA_df[IKEA_df['designer_len'] > 50].head(2)

item_id       name category  price     old_price  sellable_online  \
51  50468953  VATTVIKEN     Beds  995.0  No old price             True   
77  90331091     RIDABU     Beds  100.0        SR 125             True   

                                                 link other_colors  \
51  https://www.ikea.com/sa/en/p/vattviken-armchai...           No   
77  https://www.ikea.com/sa/en/p/ridabu-mirror-doo...           No   

                          short_description  \
51                             Armchair-bed   
77          Mirror door,          40x120 cm   

                                             designer  depth  height  width  \
51  504.689.53 Small and easy-to-place chair-bed w...   83.0    86.0   92.0   
77  903.310.91 The door can be hung with the openi...    NaN   120.0   40.0   

    designer_len  
51           403  
77           199

We can see some data in desinger is actually long description of the furniture, we can replace them with "designer unknowm", but  this might cause to misleading that those furniture are designed by the same person, so considering there are not much of them, we keep the origin ones as designer. 

In [25]:
# correlation / scatter matrix here

## Data Cleaning

In [26]:
#convert object to boolean
IKEA_df['other_colors'] = IKEA_df['other_colors'].map({'Yes': True, 'No': False})

With following observations, we can conclude that all NaN in dataset concerning the data size

In [27]:
IKEA_df[IKEA_df.isnull().any(axis=1)].shape

(1795, 14)

In [28]:
size = ['depth','height','width']
IKEA_df[IKEA_df[size].isnull().any(axis=1)].shape

(1795, 14)

In [29]:
IKEA_df[IKEA_df[IKEA_df.columns[~IKEA_df.columns.isin(size)]].isnull().any(axis=1)].shape

(0, 14)

In [30]:
IKEA_df[IKEA_df[size].isnull().all(axis=1)].head(40)

item_id                   name                    category   price  \
2     9333523  NORDVIKEN / NORDVIKEN               Bar furniture  2095.0   
9    69304221    EKEDALEN / EKEDALEN               Bar furniture  2176.0   
28   29297227    STENSELE / NORRARYD               Bar furniture  1340.0   
32   19297275    NORRÅKER / NORRÅKER               Bar furniture  1266.0   
34   20336619             HENRIKSDAL               Bar furniture    40.0   
38   29304223    EKEDALEN / EKEDALEN               Bar furniture  2375.0   
39   49304217    EKEDALEN / EKEDALEN               Bar furniture  2176.0   
42   89297272    NORRÅKER / NORRARYD               Bar furniture  1266.0   
45   49297274    NORRÅKER / NORRÅKER               Bar furniture  1266.0   
104  44361010                  BRYNE                        Beds    49.0   
117  49303393                  VIMLE                        Beds  1400.0   
151    102065         LYCKSELE LÖVÅS                        Beds   495.0   
155  60299691               BRYNILEN                        Beds    30.0   
198  60429471               DELAKTIG                        Beds   200.0   
203  45932080                 SULTAN                        Beds    50.0   
204  10309616              BÅTSFJORD                        Beds    30.0   
217  60102067         LYCKSELE HÅVET                        Beds   795.0   
222  70299695                 BJORLI                        Beds    40.0   
228  80309613               BRATTVÅG                        Beds    50.0   
231  80102066         LYCKSELE MURBO                        Beds   695.0   
406  10429987                JONAXEL  Bookcases & shelving units    15.0   
424  87749600             OBSERVATÖR  Bookcases & shelving units     5.0   
472  40347508                ENERYDA  Bookcases & shelving units    22.0   
473  80334617                   EKET  Bookcases & shelving units    12.0   
482  30284846                  BESTÅ  Bookcases & shelving units    25.0   
498  50347503                ENERYDA  Bookcases & shelving units    17.0   
503  60334312               BILLSBRO  Bookcases & shelving units    35.0   
511  10334319               BILLSBRO  Bookcases & shelving units    55.0   
529  40348715                  BESTÅ  Bookcases & shelving units    25.0   
539  80419961                JONAXEL  Bookcases & shelving units    25.0   
556    347505                ENERYDA  Bookcases & shelving units    35.0   
558    348717                  BESTÅ  Bookcases & shelving units    25.0   
615  80261258                  BESTÅ  Bookcases & shelving units    25.0   
617  70323600               BILLSBRO  Bookcases & shelving units    19.0   
634    414255              VASSVIKEN  Bookcases & shelving units   150.0   
651  60323592               BILLSBRO  Bookcases & shelving units    15.0   
653  50334303               BILLSBRO  Bookcases & shelving units    25.0   
670  50334317               BILLSBRO  Bookcases & shelving units    45.0   
676  10323599               BILLSBRO  Bookcases & shelving units    69.0   
679  20330599                UPPLEVA  Bookcases & shelving units   199.0   

        old_price  sellable_online  \
2    No old price            False   
9        SR 2,375             True   
28   No old price             True   
32       SR 1,385             True   
34   No old price             True   
38   No old price             True   
39       SR 2,375             True   
42       SR 1,385             True   
45       SR 1,385             True   
104  No old price             True   
117  No old price             True   
151  No old price             True   
155  SR 50/4 pack             True   
198  No old price             True   
203  No old price             True   
204  SR 50/4 pack             True   
217  No old price             True   
222  SR 50/4 pack             True   
228  No old price             True   
231  No old price             True   
406  No old price             True   
424  No old price             True   
472  No old pri

### NaN in size data

According to our observation to the dataset, compared with the product size shown in the website, we have the following conclusions. There are several main reasons why size data is missing \.
 First apparently some data are not successfully scraped from the website and neither be recorded in the short description, it is impossible to recover those. \
Second, some items are a set of different furniture, so there is no product size. \
Third, some items have variable size, for example, its width can be 190 to 220 cm. Although those data are recorded in short description, its hard to extract the exact size for all those items. The biggest problem is that we can’t tell to which of three, depth, height and width , the number belongs, since often that data is incomplete in short description.\
The last, there are some small items which measured in mm instead of cm and instead of height, the diameter is measured and recorded in short description.


In [31]:
#split description and size data in feature short description
IKEA_df['short_description'].str.rsplit(',', n=1,expand = True)
IKEA_df['size']= IKEA_df['short_description'].str.rsplit(',', n=1).str[1]
IKEA_df['short_description']= IKEA_df['short_description'].str.rsplit(',', n=1).str[0]

In [32]:
#split numbers and unit
IKEA_df['unit'] = IKEA_df['size'].str.replace('\d+|-|x', ' ')
IKEA_df['size_nounit'] = IKEA_df['size'].str.replace('cm|mm', '')

Following result shows as we find that indeed all small items don’t have height and considering it is convenient to extract the single numbers in short description representing the diameter, we fill the NaN in height of small items with data recorded in short description. Furthermore, we find that size data of small items, of which the diameter is measured with mm, is actually in mm according to the link, therefore we correct those.  

In [33]:
IKEA_df[IKEA_df['unit'].str.contains('mm') & IKEA_df[size].isnull().any(axis=1)].shape

(132, 17)

In [34]:
IKEA_df[IKEA_df['unit'].str.contains('mm') & IKEA_df['height'].isnull()].shape

(132, 17)

In [35]:
IKEA_df['height'] = np.where(IKEA_df['unit'].str.contains('mm') , IKEA_df.size_nounit, IKEA_df.height)
IKEA_df['height'] = np.where(IKEA_df['unit'].str.contains('mm'), IKEA_df.height.astype(float)/10, IKEA_df.height)
IKEA_df['depth'] = np.where(IKEA_df['unit'].str.contains('mm'), IKEA_df.depth.astype(float)/10, IKEA_df.depth)
IKEA_df['width'] = np.where(IKEA_df['unit'].str.contains('mm'), IKEA_df.width.astype(float)/10, IKEA_df.depth)

In [36]:
IKEA_df

item_id                   name       category   price     old_price  \
0     90420332               FREKVENS  Bar furniture   265.0  No old price   
1       368814              NORDVIKEN  Bar furniture   995.0  No old price   
2      9333523  NORDVIKEN / NORDVIKEN  Bar furniture  2095.0  No old price   
3     80155205                   STIG  Bar furniture    69.0  No old price   
4     30180504                NORBERG  Bar furniture   225.0  No old price   
...        ...                    ...            ...     ...           ...   
3689  99157902                ELVARLI      Wardrobes   750.0        SR 820   
3690   9158152                ELVARLI      Wardrobes  1572.0      SR 1,755   
3691  59157541                ELVARLI      Wardrobes   924.0      SR 1,050   
3692  89157573                ELVARLI      Wardrobes  2745.0      SR 3,130   
3693  69157376                ELVARLI      Wardrobes  1231.0      SR 1,535   

      sellable_online                                               link  \
0                True  https://www.ikea.com/sa/en/p/frekvens-bar-tabl...   
1               False  https://www.ikea.com/sa/en/p/nordviken-bar-tab...   
2               False  https://www.ikea.com/sa/en/p/nordviken-nordvik...   
3                True  https://www.ikea.com/sa/en/p/stig-bar-stool-wi...   
4                True  https://www.ikea.com/sa/en/p/norberg-wall-moun...   
...               ...                                                ...   
3689             True  https://www.ikea.com/sa/en/p/elvarli-1-section...   
3690             True  https://www.ikea.com/sa/en/p/elvarli-2-section...   
3691             True  https://www.ikea.com/sa/en/p/elvarli-2-section...   
3692             True  https://www.ikea.com/sa/en/p/elvarli-3-section...   
3693             True  https://www.ikea.com/sa/en/p/elvarli-2-section...   

      other_colors                     short_description  \
0            False                 Bar table, in/outdoor   
1            False                             Bar table   
2            False            Bar table and 4 bar stools   
3             True               Bar stool with backrest   
4            False          Wall-mounted drop-leaf table   
...            ...                                   ...   
3689         False                             1 section   
3690         False                            2 sections   
3691         False                            2 sections   
3692         False                            3 sections   
3693         False                            2 sections   

                  designer  depth height  width  designer_len  \
0     Nicholai Wiig Hansen    NaN     99    NaN            20   
1        Francis Cayouette    NaN    105    NaN            17   
2        Francis Cayouette    NaN    NaN    NaN            17   
3            Henrik Preutz   50.0    100   50.0            13   
4           Marcus Arvonen   60.0     43   60.0            14   
...                    ...    ...    ...    ...           ...   
3689       Ehlén Johansson   50.0    NaN   50.0            15   
3690       Ehlén Johansson   50.0    NaN   50.0            15   
3691       Ehlén Johansson   50.0    NaN   50.0            15   
3692       Ehlén Johansson   50.0    NaN   50.0            15   
3693       Ehlén Johansson   50.0    NaN   50.0            15   

                             size                  unit  \
0                        51x51 cm                    cm   
1                       140x80 cm                    cm   
2                             NaN                   NaN   
3                           74 cm                    cm   
4                        74x60 cm                    cm   
...                           ...                   ...   
3689             92x51x222-350 cm                    cm   
3690            135x51x222-350 cm                    cm   
3691            175x51x222-350 cm                    cm   
3692            178x51x222-350 cm                    cm   
3693  